lib/fhir/common.js

```diff
@@ -119,6 +110,196 @@ module.exports = exports = (mongo) => {
+          promises.push(new Promise((resolve, reject) => {
+            resolveReference(property.reference, (err, data) => {
+              if (err) {
+                reject(err)
+              } else {
+                resolve(data)
+              }
+            })
+          }))
+        } else {
+          return callback(`Invalid resource reference: "${JSON.stringify(property)}"`)
+        }
+      }
+    }
+
+    evaluateProperties(resultList[includeReference[0]], includeReference.slice(1))
+
+    // resolve promises
+    Promise.all(promises)
+      .then((res) => {
+        callback(null, res)
+      }).catch((err) => {
+        callback(err)
+      })
+  }
+
+  const parseSearchExpression = (exp, callback) => {
+    const itemList = exp.split(':')
+    const itemResource = itemList[0]
+    const itemReference = itemList[1].split('.')
+
+    callback(itemResource, itemReference)
+  }
+
+  const resolveIncludeParameters = (item, results) => {
+    return new Promise((resolve, reject) => {
+      const promises = []
+
+      const itemList = item.split('.')
+      const itemResource = itemList[0]
+      const itemReference = itemList.slice(1)
+
+      results.forEach((item) => {
+        if (item['resourceType'] === itemResource) {
+          promises.push(new Promise((resolve, reject) => {
+            resolveNestedProperties(item, itemReference, (err, data) => {
+              if (err) {
+                reject(err)
+              } else {
+                resolve(data)
+              }
+            })
+          }))
+        }
+      })
+
+      Promise.all(promises)
+        .then((res) => {
+          const flattenedRes = [].concat.apply([], res)
+
+          resolve(flattenedRes)
+        })
+        .catch((err) => {
+          reject(err)
+        })
+    })
+  }
+
+  function mapSearchNameToPath (searchParameters, callback) {
+    const promises = []
+    if (searchParameters) {
+      if (Array.isArray(searchParameters)) {
+        searchParameters.forEach((parameter) => {
+          promises.push(new Promise((resolve, reject) => {
+            parseSearchExpression(parameter, (resource, reference) => {
+              let searchPath = searchParamMappings.nameToPath[resource][reference]
+
+              resolve(searchPath)
+            })
+          }))
+        })
+      } else {
+        promises.push(new Promise((resolve, reject) => {
+          parseSearchExpression(searchParameters, (resource, reference) => {
+            let searchPath = searchParamMappings.nameToPath[resource][reference]
+
+            resolve(searchPath)
+          })
+        }))
+      }
+
+      Promise.all(promises)
+        .then((res) => {
+          const flattenedRes = [].concat.apply([], res)
+          callback(null, flattenedRes)
+        })
+    } else {
+      callback('No _include search parameters supplied.')
+    }
+  }
+
+  function includeResources (includeParams, results) {
+    return new Promise((resolve, reject) => {
```

Comment:
Use chaining instead. Also exit early with [Promise.reject](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject) if you want to return a rejected promise.